src/Syntax.hs

```diff
@@ -97,6 +97,22 @@ flipScope p = adjustScope go
 addScope' :: HasScopes a => a -> Scope -> a
 addScope' = adjustScope ScopeSet.insertUniversally
 
+addScopes :: HasScopes a => a -> ScopeSet -> a
+addScopes a0 scopeSet
+  = let a1 = addUniversalScopes a0 scopeSet
+        a2 = addSpecificScopes a1 scopeSet
+    in a2
+  where
+    addUniversalScopes :: HasScopes a => a -> ScopeSet -> a
+    addUniversalScopes =
+      foldlOf (to ScopeSet.contents . _1 . folded)
+              addScope'
+
+    addSpecificScopes :: HasScopes a => a -> ScopeSet -> a
+    addSpecificScopes =
+      ifoldlOf (to ScopeSet.contents .> _2 .> ifolded <. folded)
```

Comment:
This line is quite mysterious to me, but since the code seems to work, I'll assume it's some lens magic that I need to go learn :-)

Reply:
`ifoldlOf` is an "indexed" version of `foldlOf`. Here it has this type:

    ifoldlOf :: IndexedFold Phase ScopeSet Scope
             -> (Phase -> a -> Scope -> a)
             -> a -> ScopeSet -> a

An "indexed" Fold is just like a Fold, except it also grabs an index along the way. The `.>` and `<.` point to the place where the index is grabbed, and otherwise act like `.`. So the non-indexed `Fold ScopeSet Scope` would be `to ScopeSet.contents . _2 . folded . folded`, which starts from a `ScopeSet`, uses `ScopeSet.contents` to get a `(Set Scope, Map Phase (Set Scope))`, and then dives into the `Map` and the `Set` to collect all the `Scope`s. The indexed version does the same, but for each `Scope` it found, it also remembers which `Phase` it encountered along the way.

`foldlOf` is like `foldl`, but over the elements of a Fold instead of the elements of a Foldable. That is, it threads a function of type `a -> Scope -> a` through all the `Scope`s in order to compute a final `a`. The indexed version of `foldlOf` does the same, except the function also has access to the `Phase` in addition to the `Scope`.